core/common/src/main/java/alluxio/security/authentication/AuthenticatedChannelClientDriver.java

```diff
@@ -0,0 +1,195 @@
+      }
+    } catch (Throwable t) {
+      LOG.debug("Exception while handling message for {}. Message: {}. Error: {}",
+          mChannelKey.toStringShort(), saslMessage, t);
+      // Fail blocked waiters.
+      mChannelAuthenticatedFuture.setException(t);
+      mRequestObserver.onError(AlluxioStatusException.fromThrowable(t).toGrpcStatusException());
+    }
+  }
+
+  @Override
+  public void onError(Throwable throwable) {
+    LOG.debug("Received error for {}. Error: {}", mChannelKey.toStringShort(), throwable);
+    closeAuthenticatedChannel(false);
+
+    // Fail blocked waiters.
+    mChannelAuthenticatedFuture.setException(throwable);
+  }
+
+  @Override
+  public void onCompleted() {
+    LOG.debug("Authenticated channel revoked by server for {}", mChannelKey.toStringShort());
+    closeAuthenticatedChannel(false);
+  }
+
+  /**
+   * Stops authenticated session with the server by releasing the long poll.
+   */
+  public void close() {
+    LOG.debug("Closing authentication for {}", mChannelKey.toStringShort());
+    closeAuthenticatedChannel(true);
+  }
+
+  /**
+   * @return {@code true} if the channel is still authenticated
+   */
+  public boolean isAuthenticated() {
+    return mChannelAuthenticated;
+  }
+
+  /**
+   * Starts authentication with the server and wait until completion.
+   *
+   * @param timeoutMs time to wait for authentication
+   * @throws UnauthenticatedException
+   */
+  public void startAuthenticatedChannel(long timeoutMs) throws AlluxioStatusException {
+    try {
+      LOG.debug("Initiating authentication for {}", mChannelKey.toStringShort());
+      // Send the server initial message.
+      SaslMessage.Builder initialMsg = mSaslClientHandler.handleMessage(null).toBuilder();
+      initialMsg.setClientId(mChannelKey.getChannelId().toString());
+      initialMsg.setChannelRef(mChannelKey.toStringShort());
+      mRequestObserver.onNext(initialMsg.build());
+
+      // Utility to return from start when channel is secured.
+      waitUntilChannelAuthenticated(timeoutMs);
+    } catch (Throwable t) {
+      closeAuthenticatedChannel(true);
+      throw AlluxioStatusException.fromThrowable(t);
+    }
+  }
+
+  private void waitUntilChannelAuthenticated(long timeoutMs) throws AlluxioStatusException {
+    try {
+      // Wait until authentication status changes.
+      mChannelAuthenticatedFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
+      mChannelAuthenticated = true;
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+      throw AlluxioStatusException.fromThrowable(ie);
+    } catch (ExecutionException e) {
+      AlluxioStatusException statExc = AlluxioStatusException.fromThrowable(e.getCause());
+      // Unimplemented is returned if server doesn't provide authentication service.
+      if (statExc.getStatusCode() == Status.Code.UNIMPLEMENTED) {
+        throw new UnauthenticatedException("Authentication is disabled on target server.");
```

Comment:
should we include the channel info in the message?

Reply:
This exception will be wrapped by caller with channel info. Including here will double the info.